lib/omniauth/strategies/shopify.rb

```diff
@@ -49,8 +50,18 @@ def valid_signature?
       def valid_scope?(token)
         params = options.authorize_params.merge(options_for("authorize"))
         return false unless token && params[:scope] && token['scope']
-        expected_scope = params[:scope].split(',').map(&:strip).reject(&:empty?).uniq.sort
-        (expected_scope == token['scope'].split(',').sort)
+        expected_scope = normalized_scopes(params[:scope]).sort
+        (expected_scope == token['scope'].split(SCOPE_DELIMITER).sort)
+      end
+
+      def normalized_scopes(scopes)
+        scope_list = scopes.to_s.split(SCOPE_DELIMITER).map(&:strip).reject(&:empty?).uniq
+        write_scopes = scope_list.select { |scope| scope =~ /\Awrite_/ }
+
+        write_scopes.reduce(scope_list) do |remaining_scopes, write_scope|
+          read_scope = write_scope.gsub(/\Awrite_/, 'read_')
+          remaining_scopes.reject { |scope| scope == read_scope }
+        end
```

Comment:
I think this could be simplified with

```ruby
ignore_scopes = scope_list.select { |scope| scope =~ /\Awrite_/ }.map { |scope| scope.sub(/\Awrite_/, 'read_') }
scope_list - ignore_scopes
```